File: srcRuby/G01/p002_even_fibonacci_numbers.rb
# Strategy: Simple mathematics.

def sum_of_even_fibonacci_numbers(upper_limit)
  even_fibonacci_numbers_sum = 2
  fibo1 = 1
  fibo2 = 2
  next_fibo = 3

  # Use a loop since we do not know the term
  # of the Fibonacci number we are seeking.
  while next_fibo < upper_limit do
    if is_even(next_fibo) then
       even_fibonacci_numbers_sum += next_fibo
    end

    fibo1 = fibo2
    fibo2 = next_fibo
    next_fibo = fibo1 + fibo2
  end

  even_fibonacci_numbers_sum
end

def is_even(n)
  return n % 2 == 0
end

puts "Sum of even fibonacci numbers = #{sum_of_even_fibonacci_numbers(4_000_000)}"

File: src/main/java/com/translationdata/p000/P002_EvenFibonacciNumbers.java
package com.translationdata.p000;

import static org.hamcrest.CoreMatchers.*;
import static org.hamcrest.MatcherAssert.assertThat;

import org.junit.Test;

import com.translationdata.JUnitTests.FastTest;
import org.junit.experimental.categories.Category;

/** Strategy: Simple mathematics. */
@Category(FastTest.class)
public class P002_EvenFibonacciNumbers {
	int fiboSum(int upperLimit) {
		return fiboSumImpl(1, 2, 2, upperLimit);
	}

	int fiboSumImpl(final int fibo1, final int fibo2, final int total, final int upperLimit) {
		final int nextFibo = fibo1 + fibo2;
		if ( nextFibo < upperLimit) {
			if(nextFibo % 2 == 0)
				return fiboSumImpl(fibo2, nextFibo, total + nextFibo, upperLimit);
			else
				return fiboSumImpl(fibo2, nextFibo, total, upperLimit);
		}
		else
			return total;
	}

	@Test(timeout=500)
	public void SumOfEvenTermsLessThanFourMillion() {
		int result = fiboSum(4_000_000);
		System.out.printf("fiboSum(4_000_000) = %s%n", result);
		assertThat( "Incorrect sum of even Fibo numbers that do not exceed 4 million", result, is(4613732) );
	}
}
